### ISSUE TITLE
Implement Requires-Python

### ISSUE THREAD

**Issue by bennuttall** [piwheels/piwheels#208]
As suggested by @pganssle in https://github.com/pypa/setuptools/issues/1976, we should implement the `Requires-Python` attribute specified in [PEP-503](https://www.python.org/dev/peps/pep-0503/#id3) and [PEP-345](https://www.python.org/dev/peps/pep-0345/#requires-python)

It specifies:

> A repository MAY include a data-requires-python attribute on a file link. This exposes the Requires-Python metadata field, specified in PEP 345, for the corresponding release. Where this is present, installer tools SHOULD ignore the download when installing to a Python version that doesn't satisfy the requirement. For example:

```
<a href="..." data-requires-python="&gt;=3">...</a>
```

A real life example can be seen in recent versions of setuptools on PyPI:

```html
<a href="https://files.pythonhosted.org/packages/af/e7/02db816dc88c598281bacebbb7ccf2c9f1a6164942e88f1a0fded8643659/setuptools-45.0.0-py2.py3-none-any.whl#sha256=001c474c175697a03e2afffd87d92411c89215be5f2cd3e0ab80a67726c0f4c2" data-requires-python="&gt;=3.5">setuptools-45.0.0-py2.py3-none-any.whl</a><br/>
```

pip will refer to the `data-requires-python` before proceeding. This eliminates issues like https://github.com/piwheels/packages/issues/74

It would be easy enough to write this out to the simple index, but I'm not sure at first glance how we parse this information from the sdist file. Does it need to be a separate process after building the wheel?

**Comment by pganssle** [comment]
> It would be easy enough to write this out to the simple index, but I'm not sure at first glance how we parse this information from the sdist file. Does it need to be a separate process after building the wheel?

It should be available in the `PKG-INFO` file in the sdist, or in the `.dist-info` of the wheel.

I am not really familiar with how piwheels.org works, but if you are mirroring packages from PyPI, you can pull the information directly from PyPI's API listing. If you are accepting package uploads from `twine`, I believe `twine` sends the information in one of the HTTP headers.

**Comment by bennuttall** [comment]
> I am not really familiar with how piwheels.org works

We run `pip3 wheel package==version` for every package/version, on multiple distro versions (which means cp34m, cp35m and cp37m).

> but if you are mirroring packages from PyPI, you can pull the information directly from PyPI's API listing.

That sounds promising. We're just adding a utility to bring in the package short description using the JSON API. Just checked, and I see it's there. Great.

> If you are accepting package uploads from twine, I believe twine sends the information in one of the HTTP headers.

Nope, just from PyPI.

Thanks :+1:

**Comment by bennuttall** [comment]
Actually, `requires_python` in the JSON only shows in the wheel files, not the sdist, which just shows `source`. And the only place I can see it in the wheel contents is the `METADATA` file, but it's not especially parseable, beyond grep or other search. I doubt we can grab it from `PKG-INFO` as we don't explicitly download the sdist, just tell pip to build the wheel.

**Comment by pganssle** [comment]
> Actually, `requires_python` in the JSON only shows in the wheel files, not the sdist, which just shows `source`. And the only place I can see it in the wheel contents is the `METADATA` file, but it's not especially parseable, beyond grep or other search. I doubt we can grab it from `PKG-INFO` as we don't explicitly download the sdist, just tell pip to build the wheel.

If you have the `wheel` you should parse it from the `METADATA`, that's what `pip` is doing anyway. For this particular case I think it's pretty easy to parse, here's what it says in PEP 314:

> The PKG-INFO file format is a single set of RFC-822 headers parseable by the `rfc822.py` module. The field names listed in the following section are used as the header names.

That is very out of date (`rfc882` was removed from the standard library), but `setuptools` parses these files using [`email.message_from_file`](https://docs.python.org/3/library/email.parser.html#email.message_from_file). Should be pretty simple to parse the METADATA file and pull the `'Requires-Python'` field from it.

**Comment by pganssle** [comment]
> Actually, `requires_python` in the JSON only shows in the wheel files, not the sdist, which just shows `source`. And the only place I can see it in the wheel contents is the `METADATA` file, but it's not especially parseable, beyond grep or other search. I doubt we can grab it from `PKG-INFO` as we don't explicitly download the sdist, just tell pip to build the wheel.

If you have the `wheel` you should parse it from the `METADATA`, that's what `pip` is doing anyway. For this particular case I think it's pretty easy to parse, here's what it says in PEP 314:

> The PKG-INFO file format is a single set of RFC-822 headers parseable by the `rfc822.py` module. The field names listed in the following section are used as the header names.

That is very out of date (`rfc882` was removed from the standard library), but `setuptools` parses these files using [`email.message_from_file`](https://docs.python.org/3/library/email.parser.html#email.message_from_file). Should be pretty simple to parse the METADATA file and pull the `'Requires-Python'` field from it.

**Comment by brainwane** [comment]
I'm going to be pointing to this issue elsewhere to help users who are having trouble with this issue. Quoting @webknjaz in https://github.com/pypa/packaging.python.org/pull/691#discussion_r374105598 regarding interim advice:

> I'd maybe go as far as pointing to pip-tools as a workaround mentioning that this will help end-users pin versions of transitive deps.

**Comment by bennuttall** [comment]
This has just caused another issue here: https://github.com/piwheels/packages/issues/85

**Comment by bennuttall** [comment]
@waveform80 a temporary fix for this could be to add a database field for Requires-Python, I guess in the versions table or the files table. We could easily add support for this in the simple index html file, even if we have to populate the field manually for now. Would mean when we get a report like the links in the previous comment, I could add the requirement to the database and rebuild the index, rather than delete and skip. We could deploy this change along with JSON and stats and stuff without adding too much work.

**Comment by MichaIng** [comment]
Caused issue with `setuptools` earlier and with `Markdown` now: https://github.com/MichaIng/DietPi/issues/3346
For reference: https://github.com/pypa/pip/issues/7842
Really, having these flags is a must-have, otherwise users run into those kind of failures randomly when new Python2-incompatible pre-build wheels are uploaded. I have not much experience or insights but of course another issue would be to have piwheels removed from indexes completely when Python2 invokes pip, if this is somehow possible.

**Comment by bennuttall** [comment]
I've removed the latest wheels for Markdown, and opened a PR fixing the issue on their side. We're not really ready to implement this just yet, it the middle of other things and otherwise unavailable.

**Comment by MichaIng** [comment]
@bennuttall 
Okay, many thanks for the short-term fix. I see now that there are some more layers which prevent such incompatible installs usually.

**Comment by bennuttall** [comment]
This is now implemented in PR #254, pending review (and blocked by two other PRs).
